File: solutions/javascript/RemoveNthNodeFromEndofList/RemoveNthNodeFromEndofList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    var nh = new ListNode(0);
    nh.next = head;
    var len = 0;
    while (head) {
        len++;
        head = head.next;
    }
    if (len < n || n < 1) {
        return nh.next;
    }
    head = nh;
    len -= n;
    while (len) {
        len--;
        head = head.next;
    }
    head.next = head.next.next;
    return nh.next;
};

File: solutions/java/RemoveNthNodeFromEndofList/RemoveNthNodeFromEndofList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode tail = head;
        int targetNodeCount = 0;
        int totalNodes = 0;
        ListNode result = new ListNode(0);
        result.next = head;
        ListNode currentNode = result;
        if (null == head) {
            return null;
        }
        while (null != tail) {
            tail = tail.next;
            totalNodes++;
        }
        targetNodeCount = totalNodes - n;
        while (targetNodeCount > 0) {
            currentNode = currentNode.next;
            targetNodeCount--;
        }
        currentNode.next = currentNode.next.next;
        return result.next;
    }
}